// marko/dist/components/State.js
var extend = require("raptor-util/extend");

function ensure(state, propertyName) {
    var proto = state.constructor.prototype;
    if (!(propertyName in proto)) {
        Object.defineProperty(proto, propertyName, {
            get: function () {
                return this.U_[propertyName];
            },
            set: function (value) {
                this.E_(propertyName, value, false /* ensure:false */);
            }
        });
    }
}

function State(component) {
    this._a_ = component;
    this.U_ = {};

    this.r_ = false;
    this.J_ = null;
    this.I_ = null;
    this._E_ = null; // An object that we use to keep tracking of state properties that were forced to be dirty

    Object.seal(this);
}

State.prototype = {
    e_: function () {
        var self = this;

        self.r_ = false;
        self.J_ = null;
        self.I_ = null;
        self._E_ = null;
    },

    C_: function (newState) {
        var state = this;
        var key;

        var rawState = this.U_;

        for (key in rawState) {
            if (!(key in newState)) {
                state.E_(key, undefined, false /* ensure:false */
                , false /* forceDirty:false */
                );
            }
        }

        for (key in newState) {
            state.E_(key, newState[key], true /* ensure:true */
            , false /* forceDirty:false */
            );
        }
    },
    E_: function (name, value, shouldEnsure, forceDirty) {
        var rawState = this.U_;

        if (shouldEnsure) {
            ensure(this, name);
        }

        if (forceDirty) {
            var forcedDirtyState = this._E_ || (this._E_ = {});
            forcedDirtyState[name] = true;
        } else if (rawState[name] === value) {
            return;
        }

        if (!this.r_) {
            // This is the first time we are modifying the component state
            // so introduce some properties to do some tracking of
            // changes to the state
            this.r_ = true; // Mark the component state as dirty (i.e. modified)
            this.J_ = rawState;
            this.U_ = rawState = extend({}, rawState);
            this.I_ = {};
            this._a_.D_();
        }

        this.I_[name] = value;

        if (value === undefined) {
            // Don't store state properties with an undefined or null value
            delete rawState[name];
        } else {
            // Otherwise, store the new value in the component state
            rawState[name] = value;
        }
    },
    toJSON: function () {
        return this.U_;
    }
};

module.exports = State;
